core/server/common/src/main/java/alluxio/master/audit/AsyncUserAccessAuditLogWriter.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.audit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Preconditions;
+
+import java.util.concurrent.LinkedBlockingQueue;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * {@link AsyncUserAccessAuditLogWriter} writes user access audit log entries asynchronously.
+ *
+ * TODO(yanqin) investigate other lock-free queues, e.g. Lmax disruptor used by Lmax and log4j 2
+ */
+@ThreadSafe
+public final class AsyncUserAccessAuditLogWriter {
+  private static final int QUEUE_SIZE = 10000;
```

Comment:
It seems the pace at which producers produce audit log entries is slower than the consumer's consuming pace since producers have to perform a series of complex FS operations before they can append. Hopefully the producer can have enough slack. I haven't measured. As I said, it is possible that the queue is fully in certain workloads. If it does happen, the calling thread will block.
If this happens frequently, we should either increase queue capacity or look at lock-free queue.